controllers/fencer/reconciler.go

```diff
@@ -0,0 +1,206 @@
+	}
+
+	// Create an index on the VolumeAttachment's node name.
+	if err := mgr.GetFieldIndexer().IndexField(ctx, &storagev1.VolumeAttachment{}, "spec.nodeName", func(rawObj client.Object) []string {
+		va := rawObj.(*storagev1.VolumeAttachment)
+		return []string{va.Spec.NodeName}
+	}); err != nil {
+		return err
+	}
+
+	return builder.ControllerManagedBy(mgr).
+		Named("fencer").
+		WithOptions(controller.Options{MaxConcurrentReconciles: workers}).
+		WithSource(src).
+		WithEventHandler(eventHandler).
+		Complete(r)
+}
+
+// pollNodes periodically retrieves the StorageOS nodes and sends them over the
+// source event channel.  It is blocking and is intended to run in a goroutine,
+// stopping when the context is cancelled.
+func (r *Reconciler) pollNodes(ctx context.Context, src chan event.GenericEvent) {
+	ticker := time.NewTicker(r.pollInterval)
+	defer ticker.Stop()
+
+	for {
+		select {
+		case <-ticker.C:
+			tr := otel.Tracer("fencer")
+			ctx, span := tr.Start(ctx, "fencing poller")
+			r.log.V(5).Info("polling for external node health")
+
+			nodes, err := r.api.ListNodes(ctx)
+			if err != nil {
+				span.RecordError(errors.Wrap(err, "failed to list nodes"))
+				r.log.Error(err, "failed to list nodes")
+
+				// We should always be able to list nodes.  Reset the api client
+				// if we got an error.
+				r.apiReset <- struct{}{}
+
+				span.End()
+				continue
+			}
+			span.SetAttributes(label.Int("nodes", len(nodes)))
+			for _, node := range nodes {
+				src <- event.GenericEvent{
+					Object: node,
+				}
+			}
+			span.SetStatus(codes.Ok, "refreshed node cache")
+			span.End()
+		case <-ctx.Done():
+			return
+		}
+	}
+}
+
+// Reconcile is part of the main node health reconcile loop which aims to move
+// the current state of the cluster closer to the desired state.
+func (r *Reconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	tr := otel.Tracer("fencer")
+	ctx, span := tr.Start(ctx, "node fencer reconcile")
+	span.SetAttributes(label.String("name", req.NamespacedName.Name))
+	defer span.End()
+	log := r.log.WithValues("name", req.NamespacedName)
+
+	// Query the object details from the cache and operate on it.
+	val, found := r.cache.Get(req.NamespacedName.String())
+	if !found {
+		log.Error(ErrNodeNotCached, "skipping node, will retry")
+		span.RecordError(ErrNodeNotCached)
+		return ctrl.Result{}, nil
+	}
+	node := &storageosv1.Node{}
+	if err := r.scheme.Convert(val, node, nil); err != nil {
+		span.RecordError(err)
+		return ctrl.Result{}, fmt.Errorf("failed to convert cached object to Node: %v", err)
+	}
+
+	// Ignore healthy nodes
+	if node.Status.Health != storageosv1.NodeHealthOffline {
+		log.V(5).Info("ignore healthy node")
+		span.SetStatus(codes.Ok, "node healthy")
+		return ctrl.Result{}, nil
+	}
+
+	log.Info("fencing pods on node")
+	if err := r.fenceNode(ctx, node); err != nil {
+		span.RecordError(err)
+		return ctrl.Result{}, fmt.Errorf("failed to fence node: %v", err)
```

Comment:
In testing we found that the Pod was sometimes getting re-scheduled before the VolumeAttachments were deleted, causing errors to be logged (which would later succeed).  In the last commit I moved the Pod delete to after the VA delete, and returned an error for any VA delete failure.  This will cause the fence operation to requeue until all VAs and the Pod are successfully deleted. 

Previously it was ignoring the VA delete errors as the Pod was no longer available to list the PVCs->VAs if requeued.

This is much better behaviour as there's no point in deleting the Pod if there's still a VA present that will block it restarting.